README.txt
FLAM3 - cosmic recursive fractal flames
see the file COPYING for the license covering this software.

This is free software to render fractal flames as described on
http://flam3.com.  Flam3-animate makes animations, and flam3-render
makes still images.  Flam3-genome creates and manipulates genomes
(parameter sets).  A C library is also installed.

Note: the following instructions are written for Linux users.  Windows
users may install the cygwin package to get the "env" command or set
the envars in your windows command prompt manually.  That means
instead of a command like

    env dtime=5 prefix=foo. in=test.flame flam3-animate

say

    set dtime=5
    set prefix=foo.
    set in=test.flame
    flam3-animate

As usual, to configure, build, and install:

    ./configure
    make
    sudo make install

This package depends on development packages for libz, libpng,
libjpeg, and libxml2.

To test it, run

    flam3-render < test.flam3

and it should produce 00000.jpg and 00001.jpg, one image for each
<flame> element in the parameter file.  To make an animation run

    flam3-animate < test.flam3

and it should produce 100 files named 00000.jpg through 00099.jpg that
interpolate between the two <flame> elements.


envar           default         meaning
=====           =======         =======
prefix          (empty)         prefix names of output files with this string.
begin           j               time of first frame to render (j=first time in input file) (animate only)
end             n-1             time of last frame to render (n=last time specified in the input file) (animate only)
time            NA              time of first and last frame (ie do one frame) (animate only)
frame           NA              synonym for "time" (animate only)
in              stdin           name of input file
out             NA              name of output file (bad idea if rending more than one, use prefix instead)
template        NA              apply defaults based on this genome (genome only)
dtime           1               time between frames (animate only)
fields          0               if 1 then render fields, ie odd scanlines at time+0.5
nstrips         1               number of strips, ie render fractions of a frame at once (render only)
qs              1               quality scale, multiply quality of all frames by this
ss              1               size scale, multiply size (in pixels) of all frames by this
jpeg            NA              jpeg quality for compression, default is native jpeg default
format          png             "jpg" or "ppm" or "png"
pixel_aspect    1.0             aspect ratio of pixels (width over height), eg 0.90909 for NTSC
isaac_seed      random          string to be used in generating random seed.  defaults to time(0)
seed            random          integer seed for random numbers, defaults to time+pid.  deprecated.
nthreads        auto            number of threads to use (render and animate)
verbose         0               if non-zero then print progress meter on stderr
bits            33              also 16, 32, or 64: sets bit-width of internal buffers (33 means 32-bit floating-point)
bpc             8               bits per channel of color: only png supports 16 (render/animate)
image           filename        replace palette with png, jpg, or ppm image
use_vars        -1              comma sep list of variation #'s to use when generating a random flame (genome only)
dont_use_vars   unset           comma sep list of variation #'s to NOT use when generating a random flame. (genome only)
cross0          NA              randomly select one genome from this file to genetically cross (genome only)
cross1          NA                with one genome from this file (genome only)
method          NA              method for genetic cross: alternate, interpolate, or union. (genome only)
mutate          NA              randomly mutate a random genome from this file (genome only)
symmetry        NA              set symmetry of result. (genome only)
clone           NA              clone random flame in input (genome only)
clone_all       NA              clones all flames in file.  useful for applying template to all flames (genome only)
animate         NA              interpolates between all flames in a file, using times specified in file (genome only)
sequence        NA              360 degree rotation 'loops' times of each control point plus rotating transitions (genome only)
loops           NA              number of times to rotate each control point in sequence (genome only)
tries           50              number of tries to make to find a good genome. (genome only)
strip           NA              strip input, frame and nframes control which one. (genome only)
transparency    0               make bknd transparent, if format supports it (render/animate)
name_enable     0               use 'name' attr in <flame> to name image output if present (render only)
nick            ""              nickname to use in <edit> tags / img comments
url             ""              url to use in <edit> tags / img comments
id              ""              ID  to use in <edit> tags / img comments
comment         ""              comment string for <edit> tags (genome only)
use_mem         auto            floating point number of bytes of memory to use (render only)
noedits         unset           omit edit tags from output (genome only)
write_genome    0               write out genome associated with center of motion blur window (animate only)
print_edit_depth 0              depth to truncate <edit> tag structure.  0 prints all <edit> tags (genome only)
intpalette      unset           round palette entries for importing into older Apophysis versions (genome only)
insert_palette  unset           insert the palette into the image.
enable_jpeg_comments  1         enables comments in the jpeg header (render and animate)
enable_png_comments   1         enables comments in the png header (render and animate)

New in 2.8:

earlyclip       0               enables the early clipping of rgb values for better antialiasing and resizing
                                defaults to 0 for backwards compatibility
                                

for example:

    env dtime=5 prefix=foo. in=test.flam3 flam3-animate

means to render every 5th frame of parameter file foo.flam3, and store
the results in files named foo.XXXX.jpg.

the flam3-convert program reads from stdin the old format created by
the GIMP and writes to stdout the new xml format.

the flam3-genome program creates random parameter files. it also mutates,
rotates, and interpolates existing parameter files.  for example to
create 10 wholly new control points and render them at normal quality:

    env template=vidres.flam3 repeat=10 flam3-genome > new.flam3
    flam3-render < new.flam3

if you left out the "template=vidres.flam3" part then the size,
quality, etc parameters would be their default (small) values.  you
can set the symmetry group:

    env template=vidres.flam3 symmetry=3 flam3-genome > new3.flam3
    env template=vidres.flam3 symmetry=-2 flam3-genome > new-2.flam3
    flam3-render < new3.flam3
    flam3-render < new-2.flam3

Mutation is done by giving an input flame file to alter:

    env template=vidres.flam3 flam3-genome > parent.flam3
    env prefix=parent. flam3-render < parent.flam3
    env template=vidres.flam3 mutate=parent.flam3 repeat=10 flam3-genome > mutation.flam3
    flam3-render < mutation.flam3

Normally one wouldn't use the same file for the template and the file
to mutate.  Crossover is handled similarly:

    env template=vidres.flam3 flam3-genome > parent0.flam3
    env prefix=parent0. flam3-render < parent0.flam3
    env template=vidres.flam3 flam3-genome > parent1.flam3
    env prefix=parent1. flam3-render < parent1.flam3
    env template=vidres.flam3 cross0=parent0.flam3 cross1=parent1.flam3 flam3-genome > crossover.flam3
    flam3-render < crossover.flam3

flam3-genome has 3 ways to produce parameter files for animation in
the style of electric sheep.  the highest level and most useful from
the command line is the sequence method.  it takes a collection of
control points and makes an animation that has each flame do fractal
rotation for 360 degrees, then make a smooth transition to the next.
for example:

    env sequence=test.flam3 nframes=20 flam3-genome > seq.flam3
    flam3-animate < seq.flam3

creates and renders a 60 frame animation.  there are two flames in
test.flam3, so the animation consists three stages: the first one
rotating, then a transition, then the second one rotating.  each stage
has 20 frames as specified on the command line.  if you want to
render only some fraction of a whole animation file, specify the begin
and end times:

    env begin=20 end=40 flam3-animate < seq.flam3

the other two methods are harder to use because they produce files that
are only good for one frame of animation.  the output consists of 3
control points, one for the time requested, one before and one after.
that allows proper motion blur.  for example:

    env template=vidres.flam3 flam3-genome > rotme.flam3
    env rotate=rotme.flam3 frame=10 nframes=20 flam3-genome > rot10.flam3
    env frame=10 flam3-animate < rot10.flam3

the file rot10.flam3 specifies the animation for just one frame, in
this case 10 out of 20 frames in the complete animation.  C1
continuous electric sheep genetic crossfades are created like this:

    env inter=test.flam3 frame=10 nframes=20 flam3-genome > inter10.flam3
    env frame=10 flam3-animate < inter10.flam3

see http://flam3.com/flame.pdf for descriptions & formulas, and
see http://electricsheep.wikispaces.com/Variations for updates.

The complete list of variations:

   0. linear
   1. sinusoidal
   2. spherical
   3. swirl
   4. horseshoe
   5. polar
   6. handkerchief
   7. heart
   8. disc
   9. spiral
  10. hyperbolic
  11. diamond
  12. ex
  13. julia
  14. bent
  15. waves
  16. fisheye
  17. popcorn
  18. exponential
  19. power
  20. cosine
  21. rings
  22. fan
  23. blob
  24. pdj
  25. fan2
  26. rings2
  27. eyefish
  28. bubble
  29. cylinder
  30. perspective
  31. noise
  32. julian
  33. juliascope
  34. blur
  35. gaussian_blur
  36. radial_blur
  37. pie
  38. ngon
  39. curl
  40. rectangles
  41. arch
  42. tangent
  43. square
  44. rays
  45. blade
  46. secant2
  47. twintrian
  48. cross
  49. disc2
  50. super_shape
  51. flower
  52. conic
  53. parabola
  54. bent2
  55. bipolar
  56. boarders
  57. butterfly
  58. cell
  59. cpow
  60. curve
  61. edisc
  62. elliptic
  63. escher
  64. foci
  65. lazysusan
  66. loonie
  67. pre_blur
  68. modulus
  69. oscilloscope
  70. polar2
  71. popcorn2
  72. scry
  73. separation
  74. split
  75. splits
  76. stripes
  77. wedge
  78. wedge_julia
  79. wedge_sph
  80. whorl
  81. waves2
  
see http://flam3.com/flame.pdf for descriptions & formulas for each of
these.  note that, by default, if a random flame is requested and neither
'use_vars' or 'dont_use_vars' are specified, the following variations are
not used: noise, blur, gaussian_blur, radial_blur, ngon, square, rays, 
and cross.

======================================

todo:  eliminate all static storage.

======================================

changelog:

03/18/09 Major upgrade :
    - 28 variations added, mostly from the sourceforge plugins pack
    - significant speed optimization of variations
    - setting earlyclip env var on commandline will enable an alternate
      calculation of color values, resulting in better antialiasing
    - 'highlight_power' flame element controls the white highlights in
      dense areas, eliminating the hue shift bug
    - Apophysis chaos and solo xform/plotmode features have been
      implemented.  solo/plotmode are promoted to floating point
      'visibility'.  visibility and chaos are both interpolatable.
    - <motion> tag now available for xforms...allows cyclic variation
      of any regular xform parameter/coef/post.

03/17/09 Added fuzz testing with zzuf to the regression tests.  'Strip'
    mode and genomes with the zoom parameter used now break into pieces
    properly.  insert_palette fixed (broken a few versions ago.)  fixed
    twintrian variation when small weights are used.  various rare
    segfaults and memory leaks fixed.  'palette_mode' attribute added to
    flame element for smoother palette interpolation in slow animations;
    possible values are 'step' and 'linear' ('step' mode is default and 
    matches previous behaviour.)  Release as 2.7.18.

11/11/08 Add error checking on the numbers in the input genome.  Do
    not exit on read errors, fail gracefully. Changed sin/cos = tan in
    the tangent variation. Added polar to list of variations that use
    the inverted linear identity. Bugfix: temporal_filter_exp was not
    set properly in template. Apply interpolation attribute in
    templates. Add/publish datarootdir with pkg-config so the palette
    file is easy to find (for qosmic). Copyright by Spotworks LLC
    instead of Spot and Erik.  Add fuzz checks to the regression
    suite.  include LNX/OSX/WIN in the version string.  In
    flam3-genome animate command, added one to last flame time, so the
    end time is inclusive.  Release as 2.7.17.

09/06/08 Added 'clone_all' to flam3_genome to allow application of 
    template to all flames in a file at once, and 'animate' to write
    a full sequence of interpolations out.  'animate' is similar to 
    'sequence' except that no control point rotation is performed.  
    Fixed non-functional 'write_genome' env var for flam3_animate.  
    Two bugs associated with interpolating from a log interpolation_type
    to a non-log interpolation_type fixed (rotate angle reduction and 
    special inverted identity).  when using flam3_rotate in 'spin_inter'
    the interp type of the first genome is now used rather than the
    current genome's interp type.  Enforced upper and lower bounds for
    xform color and upper bound for interpolated colormap values
    as "smooth" interpolation led to values out of range.  If "smooth"
    interpolation is specified for all flames in a file, the first and
    second-to-last genome is switched to "linear" with a warning.
    spatial filters with non-box-filter window functions fixed.
    Release as 2.7.16.
    

08/25/08 Added new interpolation types 'old' and 'older', for use in 
    recreating old animations.  'linear' mode now does not rotate padded
    xforms (results in prettier symmetric singularities). switched to
    using a 'padding' flag instead of a 'just initialized' flag; padding 
    flag used for implementation of 'old' and 'older' types.  
    interpolation_space now deprecated, instead use interpolation_type. 
    flam3_align is now idempotent (multiple applications do not change
    the control points.)  Default number of temporal samples bumped to 
    1000.  Removed CVS headers from source code (now using SVN).
    Default interpolation mode now log. Removed 'move' and 'split' vars.
    changes to flam3-genome: sequence mode now returns linear 
    interpolation mode for all control points except first/last of edges 
    - these cps will use the original interpolation mode;  inter and 
    rotate modes will now return padded genomes for all control points,
    all with linear interpolation specified.  instead of centering 
    sometimes reframe by golden mean plus noise. Release as 2.7.15.

07/21/08 Add configuration option for atomic-ops.  bug fix: do not
    truncate floating point palettes.  new motion blur features: add
    temporal_filter_type, can be "box" (default) or "gaussian" or
    "exp".  Temporal_filter_width and temporal_filter_exp are parms to
    it.  'blur' env var no longer used.  Small bug fix: iteration
    count depends only on the size of the output image, not the padded
    image (the gutter).  When interpolating, only do -pi/pi adjustment
    for non-asymmetric cases.  Julian/juliascope variations use the
    alternate inverted identity for interpolation (reduces wedge
    effect).  Add python script for regression and consistency

debian/changelog
flam3 (2.7.18-0ubuntu1~ppa1) intrepid; urgency=low

  * Bug fixes from upstream.

 -- Spotworks LLC <ubuntu@spotworks.com>  Wed, 18 Mar 2009 22:29:58 -0600

flam3 (2.7.18-0ubuntu1~ppa0) hardy; urgency=low

  * Bug fixes from upstream.

 -- Spotworks LLC <ubuntu@spotworks.com>  Tue, 17 Mar 2009 20:52:50 -0600

flam3 (2.7.17-0ubuntu1~ppa0) hardy; urgency=low

  * Bug fixes from upstream.

 -- Spotworks LLC <ubuntu@spotworks.com>  Wed, 12 Nov 2008 19:07:18 -0500

flam3 (2.7.16-0ubuntu1~ppa2) hardy; urgency=low

  * Fixed version number for 2.7.16.

 -- Spotworks LLC <ubuntu@spotworks.com>  Sat, 06 Sep 2008 08:05:17 -0600

flam3 (2.7.16-0ubuntu1~ppa0) hardy; urgency=low

  * More bugs fixed and features from upstream, again interpolation related.

 -- Spotworks LLC <ubuntu@spotworks.com>  Sat, 06 Sep 2008 07:27:23 -0600

flam3 (2.7.15-0ubuntu1~ppa0) hardy; urgency=low

  * Many bugs fixed and features from upstream, mostly about interpolation.

 -- Spotworks LLC <ubuntu@spotworks.com>  Thu, 28 Aug 2008 00:28:37 -0400

flam3 (2.7.14-0ubuntu1~ppa0) hardy; urgency=low

  * Many bugs fixed and features from upstream.

 -- Spotworks LLC <ubuntu@spotworks.com>  Mon, 21 Jul 2008 18:23:45 -0400

flam3 (2.7.13-0ubuntu1~ppa8) hardy; urgency=low

  * Attempt to fix erroneous O2 compile flag.

 -- Spotworks LLC <ubuntu@spotworks.com>  Fri, 27 Jun 2008 11:23:02 -0600

flam3 (2.7.13-0ubuntu1~ppa2) hardy; urgency=low

  * Initial package release.

 -- Spotworks LLC <ubuntu@spotworks.com>  Thu, 12 Jun 2008 17:16:17 -0600



flam3.h
/*
    FLAM3 - cosmic recursive fractal flames
    Copyright (C) 1992-2009 Spotworks LLC

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef flam3_included
#define flam3_included

#include <stdio.h>
#include <libxml/parser.h>
#include "isaac.h"


char *flam3_version();

#define flam3_palette_random       (-1)
#define flam3_palette_interpolated (-2)

#define flam3_defaults_on          (1)
#define flam3_defaults_off         (0)

#define flam3_name_len    64

#define flam3_print_edits  (1)
#define flam3_dont_print_edits  (0)

//typedef double flam3_palette[256][3];
typedef struct {
	double index;
	double color[4];
} flam3_palette_entry;

typedef flam3_palette_entry flam3_palette[256];

int flam3_get_palette(int palette_index, flam3_palette p, double hue_rotation);

#define flam3_variation_random (-1)
#define flam3_variation_random_fromspecified (-2)

extern char *flam3_variation_names[];

#define flam3_nvariations 82
#define flam3_nxforms     12

#define flam3_parent_fn_len     30

#define flam3_interpolation_linear 0
#define flam3_interpolation_smooth 1

#define flam3_inttype_linear 0
#define flam3_inttype_log    1
#define flam3_inttype_compat 2 /* Linear and old behaviour */
#define flam3_inttype_older  3 /* rotate padded xforms     */

#define flam3_palette_interpolation_hsv   0
#define flam3_palette_interpolation_sweep 1

#define flam3_max_action_length 10000

#define flam3_palette_mode_step   0
#define flam3_palette_mode_linear 1

#define VAR_LINEAR   0
#define VAR_SINUSOIDAL  1
#define VAR_SPHERICAL  2
#define VAR_SWIRL 3
#define VAR_HORSESHOE  4
#define VAR_POLAR 5
#define VAR_HANDKERCHIEF 6
#define VAR_HEART 7
#define VAR_DISC 8
#define VAR_SPIRAL 9
#define VAR_HYPERBOLIC 10
#define VAR_DIAMOND 11
#define VAR_EX 12
#define VAR_JULIA 13
#define VAR_BENT 14
#define VAR_WAVES 15
#define VAR_FISHEYE 16
#define VAR_POPCORN 17
#define VAR_EXPONENTIAL 18
#define VAR_POWER 19
#define VAR_COSINE 20
#define VAR_RINGS 21
#define VAR_FAN 22
#define VAR_BLOB 23
#define VAR_PDJ 24
#define VAR_FAN2 25
#define VAR_RINGS2 26
#define VAR_EYEFISH 27
#define VAR_BUBBLE 28
#define VAR_CYLINDER 29
#define VAR_PERSPECTIVE 30
#define VAR_NOISE 31
#define VAR_JULIAN 32
#define VAR_JULIASCOPE 33
#define VAR_BLUR 34
#define VAR_GAUSSIAN_BLUR 35
#define VAR_RADIAL_BLUR 36
#define VAR_PIE 37
#define VAR_NGON 38
#define VAR_CURL 39
#define VAR_RECTANGLES 40
#define VAR_ARCH 41
#define VAR_TANGENT 42
#define VAR_SQUARE 43
#define VAR_RAYS 44
#define VAR_BLADE 45
#define VAR_SECANT2 46
#define VAR_TWINTRIAN 47
#define VAR_CROSS 48
#define VAR_DISC2 49
#define VAR_SUPER_SHAPE 50
#define VAR_FLOWER 51
#define VAR_CONIC 52
#define VAR_PARABOLA 53
#define VAR_BENT2 54
#define VAR_BIPOLAR 55
#define VAR_BOARDERS 56
#define VAR_BUTTERFLY 57
#define VAR_CELL 58
#define VAR_CPOW 59
#define VAR_CURVE 60
#define VAR_EDISC 61
#define VAR_ELLIPTIC 62
#define VAR_ESCHER 63
#define VAR_FOCI 64
#define VAR_LAZYSUSAN 65
#define VAR_LOONIE 66
#define VAR_PRE_BLUR 67
#define VAR_MODULUS 68
#define VAR_OSCILLOSCOPE 69
#define VAR_POLAR2 70
#define VAR_POPCORN2 71
#define VAR_SCRY 72
#define VAR_SEPARATION 73
#define VAR_SPLIT 74
#define VAR_SPLITS 75
#define VAR_STRIPES 76
#define VAR_WEDGE 77
#define VAR_WEDGE_JULIA 78
#define VAR_WEDGE_SPH 79
#define VAR_WHORL 80
#define VAR_WAVES2 81

typedef void (*flam3_iterator)(void *, double);


typedef struct {

   char *genome;
   char *badvals;
   char *numiters;
   char *rtime;

} flam3_img_comments;

typedef struct {

   double badvals;
   long int num_iters;
   int render_seconds;
   
} stat_struct;

typedef struct {

   unsigned int width, height;
   int version;
   int id;

   /* There are 256 levels of gray to work with */
   double intensity_weight[256];
   unsigned int bin_size[256];
   unsigned int bin_offset[256];

   /* Pointer to newly allocated memory; we will be allocating */
   /* 2*w*h ushorts for this storage.  The bin offset will     */
   /* provide the starting point for a random selection from   */
   /* (bin size) ordered pairs                                 */
   unsigned short *rowcols;

} flam3_image_store;


typedef struct xform {
   double var[flam3_nvariations];   /* interp coefs between variations */
   double c[3][2];      /* the coefs to the affine part of the function */
   double post[3][2];   /* the post transform */
   double density;      /* probability that this function is chosen. 0 - 1 */
   double color;     /* color coords for this function. 0 - 1 */
   double color_speed;  /* scaling factor on color added to current iteration */
   double animate;      /* whether or not this xform rotates (in sheep) >0 means stationary */
   double visibility;   /* 0=invisible, 1=totally visible */
   double vis_adjusted; /* adjusted visibility for better transitions */
   
   int padding;/* Set to 1 for padding xforms */
   double wind[2]; /* winding numbers */

   int precalc_angles_flag;
   int precalc_atan_xy_flag;
   int precalc_atan_yx_flag;
   double has_preblur;

   /* Params for new parameterized variations */
   /* Blob */
   double blob_low;
   double blob_high;
   double blob_waves;

   /* PDJ */
   double pdj_a;
   double pdj_b;
   double pdj_c;
   double pdj_d;

   /* Fan2 */
   double fan2_x;
   double fan2_y;

   /* Rings2 */
   double rings2_val;

   /* Perspective */
   double perspective_angle;
   double perspective_dist;

   /* Julia_N */
   double juliaN_power;
   double juliaN_dist;

   /* Julia_Scope */
   double juliaScope_power;
   double juliaScope_dist;

   /* Radial_Blur */
   double radialBlur_angle;

   /* Pie */
   double pie_slices;
   double pie_rotation;
   double pie_thickness;

   /* Ngon */
   double ngon_sides;
   double ngon_power;
   double ngon_circle;
   double ngon_corners;

   /* Curl */
   double curl_c1;
   double curl_c2;

   /* Rectangles */
   double rectangles_x;
   double rectangles_y;

   /* AMW */
   double amw_amp;

   /* Disc 2 */
   double disc2_rot;
   double disc2_twist;

   /* Supershape */
   double supershape_rnd;
   double supershape_m;
   double supershape_n1;
   double supershape_n2;
   double supershape_n3;
   double supershape_holes;
   
   /* Flower */
   double flower_petals;
   double flower_holes;
   
   /* Conic */
   double conic_eccen;
   double conic_holes;
   
   /* Parabola */
   double parabola_height;
   double parabola_width;
   
   /* Bent2 */
   double bent2_x;
   double bent2_y;
   
   /* Bipolar */
   double bipolar_shift;
   
   /* Cell */
   double cell_size;
   
   /* Cpow */
   double cpow_r;
   double cpow_i;
   double cpow_power; /* int in apo */
   
   /* Curve */
   double curve_xamp,curve_yamp;
   double curve_xlength,curve_ylength;
   
   /* Escher */
   double escher_beta;
   
   /* Lazysusan */
   double lazysusan_spin;
   double lazysusan_space;
   double lazysusan_twist;
   double lazysusan_x, lazysusan_y;
   
   /* Modulus */
   double modulus_x, modulus_y;
   
   /* Oscope */
   double oscope_separation;
   double oscope_frequency;
   double oscope_amplitude;
   double oscope_damping;
   
   /* Popcorn2 */
   double popcorn2_x, popcorn2_y, popcorn2_c;
   
   /* Separation */
   double separation_x, separation_xinside;
   double separation_y, separation_yinside;
   
   /* Split */
   double split_xsize;
   double split_ysize;
   
   /* Splits */
   double splits_x,splits_y;
   
   /* Stripes */
   double stripes_space;
   double stripes_warp;
   
   /* Wedge */
   double wedge_angle, wedge_hole;
   double wedge_count, wedge_swirl;
   
   /* Wedge_Julia */
   double wedge_julia_angle;
   double wedge_julia_count;
   double wedge_julia_power;
   double wedge_julia_dist;
   
   /* Wedge_Sph */
   double wedge_sph_angle, wedge_sph_count;
   double wedge_sph_hole, wedge_sph_swirl;
   
   /* Whorl */
   double whorl_inside, whorl_outside;
   
   /* Waves2 */
   double waves2_freqx, waves2_scalex;
   double waves2_freqy, waves2_scaley;
      
   /* If perspective is used, precalculate these values */
   /* from the _angle and _dist                         */
   double persp_vsin;
   double persp_vfcos;

   /* If Julia_N is used, precalculate these values */
   double juliaN_rN;
   double juliaN_cn;

   /* If Julia_Scope is used, precalculate these values */
   double juliaScope_rN;
   double juliaScope_cn;
   
   /* if Wedge_Julia, precalculate */
   double wedgeJulia_rN;
   double wedgeJulia_cn;
   double wedgeJulia_cf;

   /* If Radial_Blur is used, precalculate these values */
   double radialBlur_spinvar;
   double radialBlur_zoomvar;

   /* Precalculate these values for waves */
   double waves_dx2;
   double waves_dy2;

   /* If disc2 is used, precalculate these values */
   double disc2_sinadd;
   double disc2_cosadd;
   double disc2_timespi;

   /* If supershape is used, precalculate these values */
   double supershape_pm_4;
   double supershape_pneg1_n1;

   int num_active_vars;
   double active_var_weights[flam3_nvariations];
   int varFunc[flam3_nvariations];
   
   int motion_freq;
   int motion_func;
   
   struct xform *motion;
   int num_motion;
   

} flam3_xform;

typedef struct {
   char flame_name[flam3_name_len+1]; /* 64 chars plus a null */
   double time;
   int interpolation;
   int interpolation_type;
   int palette_interpolation;
   int num_xforms;
   int final_xform_index;
   int final_xform_enable;
   flam3_xform *xform;
   
   /* Xaos implementation */
   double **chaos;
   int chaos_enable;
   
   int genome_index;                   /* index into source file */
   char parent_fname[flam3_parent_fn_len];   /* base filename where parent was located */
   int symmetry;                /* 0 means none */
   flam3_palette palette;
   char *input_image;           /* preview/temporary! */
   int  palette_index;
   double brightness;           /* 1.0 = normal */
   double contrast;             /* 1.0 = normal */
   double gamma;
   double highlight_power;
   int  width, height;          /* of the final image */
   int  spatial_oversample;
   double center[2];             /* of camera */
   double rot_center[2];         /* really the center */
   double rotate;                /* camera */
   double vibrancy;              /* blend between color algs (0=old,1=new) */
   double hue_rotation;          /* applies to cmap, 0-1 */
   double background[3];
   double zoom;                  /* effects ppu, sample density, scale */
   double pixels_per_unit;       /* vertically */
   double spatial_filter_radius; /* radius of spatial filter */
   int spatial_filter_select; /* selected spatial filter */
//   double (*spatial_filter_func)(double); /* spatial filter kernel function */
//   double spatial_filter_support; /* size of standard kernel for specific function */
   double sample_density;        /* samples per pixel (not bucket) */
   /* in order to motion blur more accurately we compute the logs of the
   sample density many times and average the results. */
   /* nbatches is the number of times the buckets are filtered into
   the abucket log accumulator */
   /* ntemporal_samples is the number of time steps per batch.  this many
   interpolated control points are used per batch and accumulated */
   int nbatches;
   int ntemporal_samples;

   /* Density estimation parameters for blurring low density hits */
   double estimator;             /* Filter width for bin with one hit */
   double estimator_curve;              /* Exponent on decay function ( MAX / a^(k-1) ) */
   double estimator_minimum;         /* Minimum filter width used -
                                    forces filter to be used of at least this width on all pts */

   /* XML Edit structure */
   xmlDocPtr edits;

   /* Small-gamma linearization threshold */
   double gam_lin_thresh;

   /* for cmap_interpolated hack */
   int palette_index0;
   double hue_rotation0;
   int palette_index1;
   double hue_rotation1;
   double palette_blend;

//   double motion_exp; /* Motion blur parameter that controls how the colors are scaled */

   int temporal_filter_type; /* Temporal filters */
   double temporal_filter_width, temporal_filter_exp;
   
   int palette_mode;


} flam3_genome;

typedef struct {
	int from;
	int to;
	double scalar;
} flam3_chaos_entry;

/* xform manipulation */

void flam3_add_motion_element(flam3_xform *xf);
void flam3_add_xforms(flam3_genome *cp, int num_to_add, int interp_padding, int final_flag);
void flam3_delete_xform(flam3_genome *thiscp, int idx_to_delete);
void flam3_copy(flam3_genome *dest, flam3_genome *src);
void flam3_copyx(flam3_genome *dest, flam3_genome *src, int num_std, int num_final);
void flam3_copy_params(flam3_xform *dest, flam3_xform *src, int varn);

void flam3_xform_preview(flam3_genome *cp, int xi, double range, int numvals, int depth, double *result, randctx *rc);
unsigned short* flam3_create_xform_distrib(flam3_genome *cp);
void flam3_create_chaos_distrib(flam3_genome *cp, int xi, unsigned short *xform_distrib);
int flam3_check_unity_chaos(flam3_genome *cp);
void clear_cp(flam3_genome *cp, int def_flag);

/* samples is array nsamples*4 long of x,y,color triples.
   using (samples[0], samples[1]) as starting XY point and
   (samples[2], samples[3]) as starting color coordinate,
   perform fuse iterations and throw them away, then perform
   nsamples iterations and save them in the samples array */
int flam3_iterate(flam3_genome *g, int nsamples, int fuse, double *samples,
                     unsigned short *xform_distrib, randctx *rc);

void apply_motion_parameters(flam3_xform *xf, flam3_xform *addto, double blend);

/* genomes is array ngenomes long, with times set and in ascending order.
   interpolate to the requested time and return in result */
void flam3_interpolate(flam3_genome *genomes, int ngenomes, double time, flam3_genome *result);

/* barycentric coordinates in c */
void flam3_interpolate_n(flam3_genome *result, int ncp, flam3_genome *cpi, double *c);

/* print genome to given file with extra_attributes if not NULL */
void flam3_print(FILE *f, flam3_genome *g, char *extra_attributes, int print_edits);
char *flam3_print_to_string(flam3_genome *cp);

/* ivars is a list of variations to use, or flam3_variation_random     */
/* ivars_n is the number of values in ivars to select from.            */
/* sym is either a symmetry group or 0 meaning random or no symmetry   */
/* spec_xforms specifies the number of xforms to use, setting to 0 makes the number random. */
void flam3_random(flam3_genome *g, int *ivars, int ivars_n, int sym, int spec_xforms);

/* return NULL in case of error */
flam3_genome *flam3_parse_xml2(char *s, char *fn, int default_flag, int *ncps);
flam3_genome *flam3_parse_from_file(FILE *f, char *fn, int default_flag, int *ncps);

void flam3_add_symmetry(flam3_genome *g, int sym);

void flam3_colorhist(flam3_genome *cp, int num_batches, double *hist);
void flam3_estimate_bounding_box(flam3_genome *g, double eps, int nsamples,
             double *bmin, double *bmax, randctx *rc);
void flam3_rotate(flam3_genome *g, double angle, int interp_type); /* angle in degrees */
void flam3_align(flam3_genome *dst, flam3_genome *src, int nsrc);
void establish_asymmetric_refangles(flam3_genome *cp, int ncps);

double flam3_dimension(flam3_genome *g, int ntries, int clip_to_camera);
double flam3_lyapunov(flam3_genome *g, int ntries);

void flam3_apply_template(flam3_genome *cp, flam3_genome *templ);

int flam3_count_nthreads(void);

typedef struct {
//   double         temporal_filter_radius;
   double         pixel_aspect_ratio;    /* width over height of each pixel */
   flam3_genome  *genomes;
   int            ngenomes;
   int            verbose;
   int            bits;
   int            bytes_per_channel;
   double         time;
   int            (*progress)(void *, double, int, double);
   void          *progress_parameter;
   randctx       rc;
   int           nthreads;
} flam3_frame;


#define flam3_field_both  0
#define flam3_field_even  1
#define flam3_field_odd   2

/* out is pixel array with stride of out_width.
   pixels are rgb or rgba if nchan is 3 or 4. */
void flam3_render(flam3_frame *f, void *out, int out_width, int field, int nchan, int transp, stat_struct *stats);

double flam3_render_memory_required(flam3_frame *f);


double flam3_random01();
double flam3_random11();
int flam3_random_bit();

/* ISAAC random numbers */
double flam3_random_isaac_01(randctx *);
double flam3_random_isaac_11(randctx *);
int flam3_random_isaac_bit(randctx *);

void flam3_init_frame(flam3_frame *f);

/* External memory helpers */
void *flam3_malloc(size_t size);
void flam3_free(void *ptr);

/* AE Plugin helper functions */
size_t flam3_size_flattened_genome(flam3_genome *cp);
void flam3_flatten_genome(flam3_genome *cp, void *buf);
void flam3_unflatten_genome(void *buf, flam3_genome *cp);

void flam3_srandom();

double flam3_calc_alpha(double density, double gamma, double linrange); 
void flam3_calc_newrgb(double *cbuf, double ls, double highpow, double *newrgb);

/* Spatial filter kernels */
#define flam3_gaussian_kernel 0
#define flam3_hermite_kernel 1
#define flam3_box_kernel 2
#define flam3_triangle_kernel 3
#define flam3_bell_kernel 4
#define flam3_b_spline_kernel 5
#define flam3_lanczos3_kernel 6
#define flam3_lanczos2_kernel 7
#define flam3_mitchell_kernel 8
#define flam3_blackman_kernel 9
#define flam3_catrom_kernel 10
#define flam3_hamming_kernel 11
#define flam3_hanning_kernel 12
#define flam3_quadratic_kernel 13

/* Temporal filters */
#define flam3_temporal_box 0
#define flam3_temporal_gaussian 1
#define flam3_temporal_exp 2

/* Motion function indices */
#define MOTION_SIN 1
#define MOTION_TRIANGLE 2
#define MOTION_COS 3


#endif
